corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py

```diff
@@ -0,0 +1,208 @@
+                        enliven(sub_id)
+                    else:
+                        # need to know if parent is live before -> live
+                        extensions_by_host[ref_id].add(sub_id)
+                        hosts_by_extension[sub_id].add(ref_id)
+
+            debug('next: %r all: %r', next_ids, all_ids)
+            all_ids.update(next_ids)
+
+        # owned, open, not an extension -> live
+        for case_id in open_ids:
+            if case_id not in hosts_by_extension:
+                enliven(case_id)
+
+        # open root nodes and their extensions -> live
+        # TODO store referenced_id open/closed status in case index to avoid this query
+        root_ids = [r for r in extensions_by_host if r not in hosts_by_extension]
+        debug('roots: %r', root_ids)
+        for case_id in accessor.filter_open_case_ids(root_ids):
+            enliven(case_id)
+
+        debug('live: %r', live_ids)
+
+        with timing_context("compile_response"):
+            iaccessor = PrefetchIndexCaseAccessor(accessor, indices)
+            batches = batch_cases(iaccessor, live_ids)
+            response = compile_response(timing_context, batches, restore_state)
+
+    return response
+
+
+class PrefetchIndexCaseAccessor(object):
+
+    def __init__(self, accessor, indices):
+        self.accessor = accessor
+        self.indices = indices
+
+    def get_cases(self, case_ids, **kw):
+        assert 'prefetched_indices' not in kw
+        kw['prefetched_indices'] = [ix
+            for case_id in case_ids
+            for ix in self.indices[case_id]]
+        return self.accessor.get_cases(case_ids, **kw)
+
+
+def batch_cases(accessor, case_ids):
+    def take(n, iterable):
+        # https://docs.python.org/2/library/itertools.html#recipes
+        return list(islice(iterable, n))
+
+    ids = iter(case_ids)
+    while True:
+        next_ids = take(1000, ids)
+        if not next_ids:
+            break
+        yield accessor.get_cases(next_ids)
+
+
+def compile_response(timing_context, batches, restore_state):
+    response = restore_state.restore_class()
+
+    for cases in batches:
+        with timing_context("get_commtrack_elements"):
+            response.extend(get_commtrack_elements(cases, restore_state))
+
+        with timing_context("get_case_sync_updates (%s cases)" % len(cases)):
+            updates = get_case_sync_updates(
+                restore_state.domain, cases, restore_state.last_sync_log)
+
+        with timing_context("get_xml_for_response (%s updates)" % len(updates)):
+            response.extend(item
+                for update in updates
+                for item in get_xml_for_response(update, restore_state))
+
+    return response
+
+
+def get_commtrack_elements(cases, restore_state):
+    return get_stock_payload(
+        restore_state.project,
+        restore_state.stock_settings,
+        [CaseStub(case.case_id, case.type) for case in cases],
```

Comment:
Is it required to pass in a list of `CaseStub`s here, or would it work just as well to pass in `cases` (which is a list of real cases)?